spatialdata/_core/_spatial_query.py

```diff
@@ -57,154 +187,365 @@ def __post_init__(self) -> None:
+) -> dict[str, SpatialElement]:
+    queried_elements = {}
+    for key, element in elements.items():
+        result = query_function(element, **kwargs)
+        if result is not None:
+            # query returns None if it is empty
+            queried_elements[key] = result
+    return queried_elements
+
+
+@singledispatch
+def bounding_box_query(
+    element: Union[SpatialElement, SpatialData],
+    axes: tuple[str, ...],
+    min_coordinate: ArrayLike,
+    max_coordinate: ArrayLike,
+    target_coordinate_system: str,
+) -> Optional[Union[SpatialElement, SpatialData]]:
+    raise NotImplementedError()
+
+
+@bounding_box_query.register(SpatialData)
+def _(
+    sdata: SpatialData,
+    axes: tuple[str, ...],
+    min_coordinate: ArrayLike,
+    max_coordinate: ArrayLike,
+    target_coordinate_system: str,
+) -> SpatialData:
+    from spatialdata import SpatialData
+
+    new_elements = {}
+    for element_type in ["points", "images", "labels", "shapes"]:
+        elements = getattr(sdata, element_type)
+        queried_elements = _dict_query_dispatcher(
+            elements,
+            bounding_box_query,
+            axes=axes,
+            min_coordinate=min_coordinate,
+            max_coordinate=max_coordinate,
+            target_coordinate_system=target_coordinate_system,
+        )
+        new_elements[element_type] = queried_elements
+    return SpatialData(**new_elements, table=sdata.table)
+
+
+@bounding_box_query.register(SpatialImage)
+@bounding_box_query.register(MultiscaleSpatialImage)
+def _(
+    image: SpatialImage,
+    axes: tuple[str, ...],
+    min_coordinate: ArrayLike,
+    max_coordinate: ArrayLike,
+    target_coordinate_system: str,
+) -> Optional[Union[SpatialImage, MultiscaleSpatialImage]]:
+    """
 
     Parameters
     ----------
     image
-        The image element to perform the query on.
-    request
-        The request for the query.
+    axes
+    min_coordinate
+    max_coordinate
+    target_coordinate_system
 
     Returns
     -------
-    The image contained within the specified bounding box.
+
+    Notes
+    _____
+    See https://github.com/scverse/spatialdata/pull/151 for a detailed overview of the logic of this code,
+    and for the cases the comments refer to.
+
     """
+    # for triggering validation
+    _ = BoundingBoxRequest(
+        target_coordinate_system=target_coordinate_system,
+        axes=axes,
+        min_coordinate=min_coordinate,
+        max_coordinate=max_coordinate,
+    )
     from spatialdata._core._spatialdata_ops import (
         get_transformation,
         set_transformation,
     )
 
+    # get the transformation from the element's intrinsic coordinate system to the query coordinate space
+    transform_to_query_space = get_transformation(image, to_coordinate_system=target_coordinate_system)
+    assert isinstance(transform_to_query_space, BaseTransformation)
```

Comment:
added an exception. It will not print the element name (since it is not known from the variables in the scope), but it will print the element. This exception was already present in `get_transformation()`, now it is also printing the element.